File: ios/Claudiator/ViewModels/SessionListViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
class SessionListViewModel {
    var sessions: [Session] = []
    var isLoading = false
    var errorMessage: String?
    var filter: SessionFilter = .active

    var apiClient: APIClient?
    var deviceId: String?

    enum SessionFilter: String, CaseIterable {
        case active = "Active"
        case all = "All"
    }

    init(apiClient: APIClient? = nil) {
        self.apiClient = apiClient
    }

    func refresh() async {
        guard let apiClient, let deviceId else { return }
        await refresh(apiClient: apiClient, deviceId: deviceId)
    }

    func refresh(apiClient: APIClient, deviceId: String) async {
        if sessions.isEmpty { isLoading = true }
        do {
            // For "active" filter, exclude ended sessions client-side
            let all = try await apiClient.fetchSessions(deviceId: deviceId)
            if filter == .active {
                sessions = all.filter { $0.status != "ended" }
            } else {
                sessions = all
            }
            errorMessage = nil
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }
}

File: android/app/src/main/java/com/claudiator/app/viewmodels/SessionListViewModel.kt
package com.claudiator.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.claudiator.app.models.Session
import com.claudiator.app.services.ApiClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

enum class SessionFilter(val label: String) {
    ACTIVE("Active"),
    ALL("All"),
}

data class SessionListUiState(
    val sessions: List<Session> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val filter: SessionFilter = SessionFilter.ACTIVE,
)

class SessionListViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(SessionListUiState())
    val uiState: StateFlow<SessionListUiState> = _uiState.asStateFlow()

    fun setFilter(filter: SessionFilter) { _uiState.update { it.copy(filter = filter) } }

    fun refresh(apiClient: ApiClient, deviceId: String) {
        if (_uiState.value.sessions.isEmpty()) _uiState.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            try {
                val all = apiClient.fetchSessions(deviceId)
                val filtered = if (_uiState.value.filter == SessionFilter.ACTIVE) {
                    all.filter { it.status != "ended" }
                } else all
                _uiState.update { it.copy(sessions = filtered, isLoading = false, error = null) }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }
}
